src/v/kafka/server/handlers/api_versions.cc

```diff
@@ -71,8 +71,59 @@ serialize_apis(type_list<RequestTypes...>) {
     return apis;
 }
 
+static std::vector<api_versions_response_key>
+get_supported_apis(bool is_idempotence_enabled, bool are_transactions_enabled) {
+    auto all_api = serialize_apis(request_types{});
+
+    std::vector<api_versions_response_key> filtered;
+    std::copy_if(
+      all_api.begin(),
+      all_api.end(),
+      std::back_inserter(filtered),
+      [is_idempotence_enabled,
+       are_transactions_enabled](api_versions_response_key api) {
+          if (!is_idempotence_enabled) {
+              if (api.api_key == init_producer_id_handler::api::key) {
+                  return false;
+              }
+          }
+          if (!are_transactions_enabled) {
+              if (api.api_key == add_partitions_to_txn_handler::api::key) {
+                  return false;
+              }
+              if (api.api_key == txn_offset_commit_handler::api::key) {
+                  return false;
+              }
+              if (api.api_key == add_offsets_to_txn_handler::api::key) {
+                  return false;
+              }
+              if (api.api_key == end_txn_handler::api::key) {
+                  return false;
+              }
+          }
+          return true;
+      });
+    return filtered;
+}
+
+struct APIs {
+    APIs() {
+        base = get_supported_apis(false, false);
+        idempotence = get_supported_apis(true, false);
+        transactions = get_supported_apis(true, true);
+    }
+
+    std::vector<api_versions_response_key> base;
+    std::vector<api_versions_response_key> idempotence;
+    std::vector<api_versions_response_key> transactions;
+};
+
+static APIs supported_apis;
```

Comment:
do we want to make these thread-local? 

Other than that `static thread_local APIs supported_apis;` other than that LGTM!